README.md
# playground-aoc-spec-server

Example UserPatch spectator server relay

I wanted to figure out how easy/hard it is to implement a spectator relay. It turns out it's very easy! The UP spec server streams out a short header and then the recorded game file as it is played.

This could quite easily be adapted into a relay that supports many spectators (more than UP's max of 32), spec delays, a live recorded game analysis, etc.

This example supports joining at any point after the game has started, regardless of the late join configuration in UP itself.

[Remaining Questions](#remaining-questions) - [More](#more) - [Usage](#usage) - [License: Apache-2.0](#license)

## Remaining Questions

The spec client doesn't work the same as UP's spectate.exe. When reading to the end of a rec file, spectate.exe waits for more data, but with client.js the game ends.

## More

Since it was easier than expected, I might try adding more stuff to this proof-of-concept:

 - spec delay
 - live data stream using recage, to test its streams impl
 - ???

## Usage

First start a UserPatch game that allows spectators (Sx button in the game setup room)

then do:
```bash
# provide the host IP that's runnning a spectatable game
node server 192.168.178.116
```

Then you can spec using
```bash
# provide IP of the box running the spec relay server
node client ~/path/to/Age2_x1.5.exe 192.168.178.112
```

## License

[Apache-2.0](LICENSE.md)


CHANGELOG.md
# playground-aoc-spec-server change log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).


index.js
const net = require('net')
const util = require('util')
const struct = require('awestruct')
const args = require('minimist')(process.argv[2])

const host = args._[0]
if (!host) {
  console.error('provide the IP that\'s hosting the game')
  process.exit(1)
}

// spectator info header
const t = struct.types
const readString = (s) => s.slice(0, s.indexOf(0)).toString('ascii')
const writeString = (n) => (s) => {
  const b = Buffer.alloc(n)
  b.write(s, 0, s.length, 'ascii')
  return b
}
const SpecHeader = struct([
  // C-style strings with lots of uninitialised memory past their end, so read as buffer and convert manually
  ['gameName', t.buffer(32).map(readString, writeString(32))],
  ['fileType', t.buffer(32).map(readString, writeString(32))],
  ['playerName', t.buffer(192).map(readString, writeString(192))]
])

let isFirst = false
// keep packets in memory for late join
const buffers = []
const clients = new Set()
const socket = net.createConnection(53754, '192.168.178.116', () => {
  console.log('connected to spec server')

  socket.on('data', (packet) => {
    if (isFirst) {
      console.log(SpecHeader(packet))
      isFirst = false
    }

    // A spec delay impl could check the sync packets here for the current game time, and buffer accordingly
    buffers.push(packet)
    for (const client of clients) {
      client.write(packet)
    }
  })

  socket.on('end', () => {
    for (const client of clients) {
      client.end()
    }
    clients.clear()
  })
})

// Example spectator server
net.createServer((client) => {
  console.log('new client', client.remoteAddress)
  clients.add(client)
  // Push out all the data we missed
  for (const packet of buffers) {
    client.write(packet)
  }
}).listen(53754, '127.0.0.1')


server.js
const net = require('net')
const util = require('util')
const struct = require('awestruct')
const args = require('minimist')(process.argv[2])

const host = args._[0]
if (!host) {
  console.error('provide the IP that\'s hosting the game')
  process.exit(1)
}

// spectator info header
const t = struct.types
const readString = (s) => s.slice(0, s.indexOf(0)).toString('ascii')
const writeString = (n) => (s) => {
  const b = Buffer.alloc(n)
  b.write(s, 0, s.length, 'ascii')
  return b
}
const SpecHeader = struct([
  // C-style strings with lots of uninitialised memory past their end, so read as buffer and convert manually
  ['gameName', t.buffer(32).map(readString, writeString(32))],
  ['fileType', t.buffer(32).map(readString, writeString(32))],
  ['playerName', t.buffer(192).map(readString, writeString(192))]
])

let isFirst = false
// keep packets in memory for late join
const buffers = []
const clients = new Set()
const socket = net.createConnection(53754, host, () => {
  console.log('connected to spec server')

  socket.on('data', (packet) => {
    if (isFirst) {
      console.log(SpecHeader(packet))
      isFirst = false
    }

    // A spec delay impl could check the sync packets here for the current game time, and buffer accordingly
    buffers.push(packet)
    for (const client of clients) {
      client.write(packet)
    }
  })

  socket.on('end', () => {
    for (const client of clients) {
      client.end()
    }
    clients.clear()
  })
})

// Example spectator server
net.createServer((client) => {
  console.log('new client', client.remoteAddress)
  clients.add(client)
  // Push out all the data we missed
  for (const packet of buffers) {
    client.write(packet)
  }
}).listen(53754, '127.0.0.1')


client.js
const net = require('net')
const path = require('path')
const { spawnSync, spawn } = require('child_process')
const fs = require('fs')
const struct = require('awestruct')
const args = require('minimist')(process.argv.slice(2))

const [aoc, host] = args._

if (!aoc) {
  console.error('must provide path to the aoc executable')
  process.exit(1)
}
if (!host) {
  console.error('must provide spec server hostname')
  process.exit(1)
}

// spectator info header
const t = struct.types
const readString = (s) => s.slice(0, s.indexOf(0)).toString('ascii')
const writeString = (n) => (s) => {
  const b = Buffer.alloc(n)
  b.write(s, 0, s.length, 'ascii')
  return b
}
const SpecHeader = struct([
  // C-style strings with lots of uninitialised memory past their end, so read as buffer and convert manually
  ['gameName', t.buffer(32).map(readString, writeString(32))],
  ['fileType', t.buffer(32).map(readString, writeString(32))],
  ['playerName', t.buffer(192).map(readString, writeString(192))]
])

const socket = net.createConnection(53754, host, () => {
  socket.once('data', (packet) => {
    const header = SpecHeader(packet)
    const mgx = path.join(path.dirname(aoc), `../SaveGame/spc.0.${header.playerName}.${header.fileType}`)

    console.log('spec file:', mgx)
    socket.pipe(fs.createWriteStream(mgx))
    // wait for header to be fully loaded
    socket.once('data', (header) => {
      const len = header.readUInt32LE()
      console.log(`waiting for full header... (${len} bytes)`)
      let current = header.length
      socket.on('data', function ondata (chunk) {
        current += header.length
        if (current >= len) {
          socket.off('data', ondata)
          start(header.gameName, mgx)
        }
      })
    })
  })
})

function start (game, mgx) {
  console.log('starting game')
  const recpath = spawnSync('winepath', ['-w', mgx]).stdout.toString().trim()
  console.log(recpath)
  spawn('wine', [aoc, `GAME=${game}`, recpath], { stdio: 'inherit' }).on('close', () => {
    console.log('done')
  })
}
